File: backtrack_recursion/39 combination sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        visited=[False]*len(candidates)
        self.backtrack(candidates, 0, [], target, res)
        return res
    
    def backtrack(self, candidates, index, path, target, res):
        if target == 0:
            res.append(path)
            return
        if target < 0:
            return
        for i in range(index, len(candidates)):
            self.backtrack(candidates, i, path+[candidates[i]], target-candidates[i], res)
            
        

File: backtrack_recursion/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;
        backtrack(candidates, 0, target, path, res);
        return res;
    }
    
    void backtrack(vector<int>& candidates, int index, int target, vector<int>& path, vector<vector<int>>& res)
    {
        if(path.size() != 0 && target == 0)
        {
            res.push_back(path);
            return ;
        }
        if(target < 0)
            return ;
        for(int i=index; i < candidates.size(); i++)
        {
            path.push_back(candidates[i]);
            backtrack(candidates, i, target - candidates[i], path, res);
            path.pop_back();
        }
    }
};
